opt: add exploration rules to hoist project from under join

RaduBerinde opened the pull request:
UPSERTs with virtual columns result in plans that can't utilize lookup
joins because of a Project inside a left join.

This change adds two exploration rules that try to pull up the
Project, allowing other rules to fire on the resulting join. For inner
joins, this is a trivial transformation (as long as the ON condition
doesn't refer to a projection). For the right side of left joins (the
UPSERT case), this is more complicated: we have to find a canary
column from the right side. Fortunately, for the typical `a=b` join
condition, `a` needs to be non-NULL, so we can use such a column as
canary.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59780)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
Beware that this will cause join ordering to be recomputed on more combinations. You might want to try out large join trees with projects interspersed in them and try to find any possibility of exponential explosions.

Also, how come this needs to be an exploration rule? I'd think we'd always want to pull `Project` operators up the tree, as long as they contain non pass-through cols.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-:-MSftuU31H4CP0RVKu5W:b6d5w3p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I could always disable reordering on the generated joins? The main goal of the rule is to get a lookup join.

> Also, how come this needs to be an exploration rule? I'd think we'd always want to pull Project operators up the tree, as long as they contain non pass-through cols.

The output of a join could be much larger than an input, we'd be recalculating the same projection many more times.

RaduBerinde commented:
Note that the join we create inherits the `SkipReorderJoins` flag, so if it was a result of the reordering, it won't be reordered again.

andy-kimball reviewed (commented):
I see, OK.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-:-MSkGi3MBv7C9WKrkqaF:b5o8c9z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 7 of 7 files at r1, 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1738 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSn3OP87IWB-x_EcyI6:-MSn3OP87IWB-x_EcyI7:b-4q09rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/logical_props_builder.go#L1738)):*
> ```Go
> 
> 	case opt.IsNotOp:
> 		if condition.(*IsNotExpr).Right.Op() == opt.NullOp {
> ```

Anything to consider here for `IsTupleNotNullExpr`? I think the answer is "no" because we only null reject variables that are direct children of the comparison op.
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1754 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSn4GiG1ZTPHID-6Zu2:-MSn4GiG1ZTPHID-6Zu3:bkbvel0) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/logical_props_builder.go#L1754)):*
> ```Go
> 		filterProps := filters[i].ScalarProps()
> 		if filterProps.Constraints != nil {
> 			notNullCols.UnionWith(filterProps.Constraints.ExtractNotNullCols(evalCtx))
> ```

It looks like constraints already determine that `x=y` is null rejecting. From one of your tests:

```
      └── filters
           └── eq [type=bool, outer=(1,2), constraints=(/1: (/NULL - ]; /2: (/NULL - ]), fd=(1)==(2), (2)==(1)]
                ├── variable: x:1 [type=int]
                └── variable: y:2 [type=int]
```

I'm confused why `ExtractNotNullCols` doesn't extract `x` and `y` in this case. Why is this extra logic with `ConditionRejectsNullInputs` necessary?
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1757 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSn4r1KB59a2n4MADbN:-MSn4Exm3GslXKOCUXMK:b-x1akwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/logical_props_builder.go#L1757)):*
> ```Go
> 		}
> 		// If the operator returns NULL on NULL inputs, we can null-reject all its
> 		// inputs.
> ```

[nit] maybe "... we can null-reject all its inputs that are variables"?
___
*[pkg/sql/opt/xform/join_funcs.go, line 907 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSn9y4p0WPtF199i-m7:-MSn9y4p0WPtF199i-m8:b-rszfty) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/join_funcs.go#L907)):*
> ```Go
> }
> 
> func (c *CustomFuncs) FindLeftJoinCanaryColumn(
> ```

[nit] add short explanation for what this function does.
___
*[pkg/sql/opt/xform/join_funcs.go, line 928 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnBKw03z2lFZj58H3q:-MSnBKw03z2lFZj58H3r:bcgfd87) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/join_funcs.go#L928)):*
> ```Go
> 	}
> 
> 	return 0
> ```

[nit] something to consider in the future - a `const UnknownColumnID = 0` that can be used here and in all the places we check for a zero value of a column ID.
___
*[pkg/sql/opt/xform/rules/join.opt, line 274 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnBgZmE54mmvCyQiSO:-MSnBgZmE54mmvCyQiSP:bv1pb2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/rules/join.opt#L274)):*
> ```
> # This rule is similar to HoistProjectFromInnerJoin, except that we have to
> # handle "outer" rows correctly and make sure we project NULLs, regardless of
> # the expressions.
> ```

[nit] A quick explanation of how this is done might be useful. Something like: "First we find a canary column that will only be null if there was no match on the right side of the join. Then, we wrap each projection in a CASE statement which is null if the canary column is null, and equivalent to the original projection if the canary column is not null".
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnCoBx-DN0h2XvRn5g:-MSnCoBx-DN0h2XvRn5h:b-jd0rlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7156)):*
> ```
>       └── CASE p:5 IS NULL WHEN true THEN CAST(NULL AS INT8) ELSE p:5 + q:6 END [as="?column?":11, outer=(5,6), immutable]
> 
> # The rule should use a as the canary column.  opt expect=HoistProjectFromLeftJoin
> ```

[nit] remove `opt expect=HoistProjectFromLeftJoin`
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnEEX8873uGY-eTgBJ:-MSnEEX8873uGY-eTgBK:b8mmvlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7175)):*
> ```
> # TODO(radu): we could try adding a column to the input scan, similar to EnsureKey.
> opt expect-not=HoistProjectFromLeftJoin
> SELECT * FROM small LEFT JOIN (SELECT a, a+b FROM abcd) ON a=m OR m=1
> ```

There's also the possibility of using a not-null column on the right side of the join as the canary column. For example:

```
SELECT * FROM small LEFT JOIN (SELECT a, a+b, x FROM abcd) ON a=m OR m=1
```

`x` can be the canary column if it is `NOT NULL`, right?
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnEmejC90ckRBMoS-0:-MSnEmejC90ckRBMoS-1:bebnb7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7192)):*
> ```
>       └── (a:5 = m:1) OR (m:1 = 1) [outer=(1,5), constraints=(/1: (/NULL - ])]
> 
> # The rule should not fire when a projection is volatile.
> ```

Can you add a test case similar to the one you have for `HoistProjectFromLeftJoin`: "The rule should not fire when the ON condition uses a projection."


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/opt/xform/rules/join.opt, line 274 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnBgZmE54mmvCyQiSO:-MSnFWk9EOZHC4_EdnEm:b-ib2b94) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/rules/join.opt#L274)):*
> First we find a canary column that will only be null if there was no match on the right side of the join.

Edit: "First we find a canary column that is null-rejected by the ON condition."


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 7 of 7 files at r1, 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1734 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnvXm41jcMGkN4X1GI:-MSnvXm59mYlLnL91kug:br41y0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/logical_props_builder.go#L1734)):*
> ```Go
> 		opt.RegMatchOp, opt.NotRegMatchOp, opt.RegIMatchOp, opt.NotRegIMatchOp,
> 		opt.ContainsOp, opt.JsonExistsOp, opt.JsonAllExistsOp, opt.JsonSomeExistsOp,
> 		opt.OverlapsOp, opt.BBoxCoversOp, opt.BBoxIntersectsOp:
> ```

Is there a good way we can validate this list (and make sure it stays up to date)? I guess this is kind of similar to some of our functions in `operator.go` like `AggregateIsNeverNull` (speaking of which, should this function live there?). Maybe in the meantime you could just update https://github.com/cockroachdb/cockroach/issues/55851 to mention that we'd like to include this function in any future test infrastructure.
___
*[pkg/sql/opt/memo/testdata/logprops/select, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnx89IDhP5q9GFaSpb:-MSnx89J0i8ylG8dur_z:b-v9zbur) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/testdata/logprops/select#L37)):*
> ```
>                 └── const: 1 [type=int]
> 
> # We should determine that y is not null.
> ```

[nit] the way this is written it sounds like a TODO, but it looks like you're already correctly determining that y is not null.
___
*[pkg/sql/opt/xform/join_funcs.go, line 931 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnycuw3eMKPhR0Jo1l:-MSnycux4Abtt65KSltT:b-9txb87) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/join_funcs.go#L931)):*
> ```Go
> }
> 
> func (c *CustomFuncs) FoundCanaryColumn(canaryCol opt.ColumnID) bool {
> ```

[nit] needs a comment (it's failing the lint check)


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: add exploration rules to hoist project from under join

UPSERTs with virtual columns result in plans that can't utilize lookup
joins because of a Project inside a left join.

This change adds two exploration rules that try to pull up the
Project, allowing other rules to fire on the resulting join. For inner
joins, this is a trivial transformation (as long as the ON condition
doesn't refer to a projection). For the right side of left joins (the
UPSERT case), this is more complicated: we have to find a canary
column from the right side. Fortunately, for the typical `a=b` join
condition, `a` needs to be non-NULL, so we can use such a column as
canary.

Release note: None

RaduBerinde reviewed (commented):
I added some comments about SkipReorderJoins to the rules. I removed the first commit as it turns out it was unnecessary

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-:-MT6sZxw4b_DOMfB7ifF:bhfvkg7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1754 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSn4GiG1ZTPHID-6Zu2:-MT6jA_sDoYiuobvKf9A:bak8y8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/logical_props_builder.go#L1754)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It looks like constraints already determine that `x=y` is null rejecting. From one of your tests:

```
      └── filters
           └── eq [type=bool, outer=(1,2), constraints=(/1: (/NULL - ]; /2: (/NULL - ]), fd=(1)==(2), (2)==(1)]
                ├── variable: x:1 [type=int]
                └── variable: y:2 [type=int]
```

I'm confused why `ExtractNotNullCols` doesn't extract `x` and `y` in this case. Why is this extra logic with `ConditionRejectsNullInputs` necessary?
</blockquote></details>

Hm.. I screwed up. I was looking at `EXPLAIN (OPT, VERBOSE)` which apparently doesn't show the `!null` stuff. Removed this commit altogether.
___
*[pkg/sql/opt/memo/testdata/logprops/select, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnx89IDhP5q9GFaSpb:-MT6kOT81qoQeBjmahby:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/35ff0fb9ad89da46d8a9f66e1e13c869faf941f9/pkg/sql/opt/memo/testdata/logprops/select#L37)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] the way this is written it sounds like a TODO, but it looks like you're already correctly determining that y is not null.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/join_funcs.go, line 928 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnBKw03z2lFZj58H3q:-MT6sloy5Oj4aME7VIXx:bwnxo7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/join_funcs.go#L928)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

[nit] something to consider in the future - a `const UnknownColumnID = 0` that can be used here and in all the places we check for a zero value of a column ID.
</blockquote></details>

Not sure UnknownColumnID is much better than just 0..
___
*[pkg/sql/opt/xform/join_funcs.go, line 931 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnycuw3eMKPhR0Jo1l:-MT6lS-BC97sx1xoxDFH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/join_funcs.go#L931)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] needs a comment (it's failing the lint check)
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnCoBx-DN0h2XvRn5g:-MT6qgAQ9CNBEx6ytTgq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7156)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

[nit] remove `opt expect=HoistProjectFromLeftJoin`
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnEEX8873uGY-eTgBJ:-MT6rqmzAREEziubVhST:bdmspfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7175)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

There's also the possibility of using a not-null column on the right side of the join as the canary column. For example:

```
SELECT * FROM small LEFT JOIN (SELECT a, a+b, x FROM abcd) ON a=m OR m=1
```

`x` can be the canary column if it is `NOT NULL`, right?
</blockquote></details>

With ON conditions like that, we can't turn it into a lookup join.. So the rule fires but it's not used in the resulting plan. I added a test with a simple equality condition, but it is a bit fragile in that we could always choose the equality column.
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnEmejC90ckRBMoS-0:-MT6sXz53lqg_526q6HY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7192)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a test case similar to the one you have for `HoistProjectFromLeftJoin`: "The rule should not fire when the ON condition uses a projection."
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 10 of 10 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-:-MT6z8np9qCv7GWrsJF-:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 928 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnBKw03z2lFZj58H3q:-MT6xNGb3hQ6Wvmda-u1:b-gxztsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/join_funcs.go#L928)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not sure UnknownColumnID is much better than just 0..
</blockquote></details>

Fair enough.
___
*[pkg/sql/opt/xform/testdata/rules/join, line 7175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59780#-MSnEEX8873uGY-eTgBJ:-MT6yrAYDQ5ETzIaTOqh:b9zozbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/436f836a69c865a342d3c67f7fade71218764ff1/pkg/sql/opt/xform/testdata/rules/join#L7175)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

With ON conditions like that, we can't turn it into a lookup join.. So the rule fires but it's not used in the resulting plan. I added a test with a simple equality condition, but it is a bit fragile in that we could always choose the equality column.
</blockquote></details>

Oops. Makes sense. :thumbsup: 


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
